### PR TITLE
Generate IOperation Interfaces and OperationKind enum

### PR DESCRIPTION
**Pull request by 333fred**
I've had generating IOperations on my backlog for a long time now, and as part of ripping out the laziness in IOperation I think it's a good time to finally accomplish it. This is the first in a series of PRs to accomplish this, generating _just_ the IOperation interfaces themselves and the OperationKind enum.

I highly, highly, highly, super highly, seriously recommend reviewing this commit by commit:
* Commit 1 introduces the generation structure and generates the interfaces and enum _in place_. This allows for easy diffing of the changes that the generator makes compared to the existing definitions. By an large it's just whitespace, with a couple of small comment changes. No binary changes were introduced.
* Commit 2 moves the generated output from being in the original files to being in 3 files, one for each namespace we're generating. It also enables verification like we have for the bound nodes, so if changes are made to the files or the xml we'll get a CI error.

@dotnet/roslyn-compiler for review. /cc @mavasani @genlu.

### CODE REVIEW

File: Roslyn.sln [hunk 1]
```diff
@@ -452,689 +454,1115 @@ Global
 	EndGlobalSection
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
 		Debug|Any CPU = Debug|Any CPU
+		Debug|x64 = Debug|x64
```
Comment (jasonmalinowski) [inline Roslyn.sln]: I suspect you didn't intend to have this happen (and it'll probably cause funky things too.)

**Comment by genlu** [comment]
> ... and as part of ripping out the laziness in IOperation...

Just curious, what's this about?

**Comment by 333fred** [comment]
@ge

File: src/Compilers/Core/Portable/Operations/IConstructorBodyOperation.cs [hunk 2]
```diff
@@ -18,12 +18,11 @@ namespace Microsoft.CodeAnalysis.Operations
     public interface IConstructorBodyOperation : IMethodBodyBaseOperation
     {
         /// <summary>
-        /// Local declarations contained within the <see cref="Initializer"/>.
+        /// Local declarations contained within the <see cref="Initializer" />.
```
Comment (genlu) [inline src/Compilers/Core/Portable/Operations/IConstructorBodyOperation.cs]: Hmm, did we have a whitespace in the file name before?

File: Roslyn.sln [hunk 3]
```diff
@@ -426,6 +426,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.VisualStudio.Lang
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "IdeBenchmarks", "src\Tools\IdeBenchmarks\IdeBenchmarks.csproj", "{FF38E9C9-7A25-44F0-B2C4-24C9BFD6A8F6}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "CompilersIOperationGenerator", "src\Tools\Source\CompilerGeneratorTools\Source\IOperationGenerator\CompilersIOperationGenerator.csproj", "{D55FB2BD-CC9E-454B-9654-94AF5D910BF7}"
```
Comment (genlu) [inline Roslyn.sln]: Not sure it this matters, but should this be part of the src\Tools\Source\CompilerGeneratorTools\CompilerTools.sln as well?

File: src/Compilers/Core/Portable/Operations/IConstructorBodyOperation.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/IConstructorBodyOperation.cs]: Yes. Yes we did.

Comment (genlu) [reply src/Compilers/Core/Portable/Operations/IConstructorBodyOperation.cs]: 🤦‍♂️

File: src/Compilers/Core/Portable/Operations/PlaceholderKind.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+namespace Microsoft.CodeAnalysis.Operations
+{
+    internal enum PlaceholderKind
+    {
+        Unspecified = 0,
+        SwitchOperationExpression = 1,
+        ForToLoopBinaryOperatorLeftOperand = 2,
+        ForToLoopBinaryOperatorRightOperand = 3,
+        AggregationGroup = 4,
+    }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Operations/PlaceholderKind.cs]: What is this for?

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml [hunk 5]
```diff
@@ -0,0 +1,2862 @@
...
+      UnusedOperationKinds indicates kinds that are currently skipped by the OperationKind
+      enum. They can be used by future nodes by inserting those nodes at the correct point
+      in the list.
+        -->
+
+  <UnusedOperationKinds>
+    <Entry Value="0x1D"/>
+    <Entry Value="0x62"/>
+    <Entry Value="0x64"/>
+  </UnusedOperationKinds>
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: what does this mean?

**Comment by CyrusNajmabadi** [comment]
NO real concerns.  Glad to see this happen :)

File: src/Compilers/Core/Portable/Operations/PlaceholderKind.cs (hunk 4 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/PlaceholderKind.cs]: It's used by IPlaceholderOperation. Previously this was defined in the same file as the operation itself, so I had to split it out.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Operations/PlaceholderKind.cs]: not the responsibility of the generator to create this?

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml (hunk 5 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: These were skipped for various reasons, usually due to changing apis that meant something wasn't shipped. I had to skip them in the OperationKind enum though, so they're enumerated here, and future operations could make use of these kinds.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler please review.

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+            {
+                var @namespace = grouping.Key ?? "Operations";
+                var outFileName = Path.Combine(_location, $"{@namespace}.Generated.cs");
+                using (_writer = new StreamWriter(File.Open(outFileName, FileMode.Create), Encoding.UTF8))
+                {
+                    writeHeader();
+                    WriteUsing("System");
+                    WriteUsing("System.Collections.Immutable");
+
+                    if (grouping.Key is object)
```
Comment (RikkiGibson) [inline src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs]: >if (grouping.Key is object) [](start = 20, length = 27)

I guess it doesn't matter for an internal tool but will this include an unnecessary using if the XML contains an explicit "Operations" namespace?

Comment (333fred) [reply src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs]: Yes, we would. I could add assert that it doesn't.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review please

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review please

File: src/Compilers/Core/Portable/Operations/ICoalesceOperation.cs [hunk 7]
```diff
@@ -1,11 +1,12 @@
...
+using System;
 
 namespace Microsoft.CodeAnalysis.Operations
 {
     /// <summary>
     /// Represents a coalesce operation with two operands:
-    ///  (1) <see cref="Value"/>, which is the first operand that is unconditionally evaluated and is the result of the operation if non null.
-    ///  (2) <see cref="WhenNull"/>, which is the second operand that is conditionally evaluated and is the result of the operation iff <see cref="Value"/> is null.
+    /// (1) <see cref="Value" />, which is the first operand that is unconditionally evaluated and is the result of the operation if non null.
+    /// (2) <see cref="WhenNull" />, which is the second operand that is conditionally evaluated and is the result of the operation if <see cref="Value" /> is null.
```
Comment (agocke) [inline src/Compilers/Core/Portable/Operations/ICoalesceOperation.cs]: `iff` looks right to me

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml [hunk 8]
```diff
@@ -0,0 +1,2862 @@
+﻿<?xml version="1.0" encoding="UTF-8"?>
+<!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
+<Tree Root="IOperation">
+
+
+  <!--
+      To regenerate the operation nodes, run eng/generate-compiler-code.cmd
+
+      The operations in this file are _ordered_! If you change the order in here,
+      you will affect the ordering in the generated OperationKind enum, which will
```
Comment (agocke) [inline src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: Hm, interesting.

File: src/Compilers/Core/Portable/Operations/OperationKind.cs [hunk 9]
```diff
@@ -22,7 +20,7 @@ public enum OperationKind
         VariableDeclarationGroup = 0x3,
```
Comment (agocke) [inline src/Compilers/Core/Portable/Operations/OperationKind.cs]: FWIW, I don't find the hex particularly useful in this class. It doesn't really matter what the enum bitpattern is, so the hex doesn't help me. Not something to change, just an observation.

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+
+                    string[] separators = new[] { "\r", "\n", "\r\n" };
+                    string[] lines = el.InnerXml.Split(separators, StringSplitOptions.RemoveEmptyEntries);
+
+                    int indentation = lines[0].Length - lines[0].TrimStart().Length;
+
+                    foreach (var line in lines)
+                    {
+                        if (string.IsNullOrWhiteSpace(line)) continue;
+                        WriteLine($"/// {line.Substring(indentation)}");
```
Comment (agocke) [inline src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs]: Thank you for using interpolated strings in this generator, the other ones are borderline unreadable without it.

File: Roslyn.sln (hunk 3 above)
Comment (333fred) [reply Roslyn.sln]: I don't think anyone uses that sln, if they need to update the project and want to use that sln they can update at that time.

---
In reply to: [309433193](https://github.com/dotnet/roslyn/pull/37624#discussion_r309433193) [](ancestors = 309433193)

**Pull request merged**